executor/mem_reader.go

```diff
@@ -0,0 +1,428 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/distsql"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+)
+
+type memIndexReader struct {
+	ctx           sessionctx.Context
+	index         *model.IndexInfo
+	table         *model.TableInfo
+	kvRanges      []kv.KeyRange
+	desc          bool
+	conditions    []expression.Expression
+	addedRows     [][]types.Datum
+	retFieldTypes []*types.FieldType
+	outputOffset  []int
+	// cache for decode handle.
+	handleBytes []byte
+	// memIdxHandles is uses to store the handle ids that has been read by memIndexReader.
+	memIdxHandles map[int64]struct{}
+	// belowHandleIndex is the handle's position of the below scan plan.
+	belowHandleIndex int
+}
+
+func buildMemIndexReader(us *UnionScanExec, idxReader *IndexReaderExecutor) *memIndexReader {
+	kvRanges := idxReader.kvRanges
+	outputOffset := make([]int, 0, len(us.columns))
+	for _, col := range idxReader.outputColumns {
+		outputOffset = append(outputOffset, col.Index)
+	}
+	return &memIndexReader{
+		ctx:              us.ctx,
+		index:            idxReader.index,
+		table:            idxReader.table.Meta(),
+		kvRanges:         kvRanges,
+		desc:             us.desc,
+		conditions:       us.conditions,
+		addedRows:        make([][]types.Datum, 0, len(us.dirty.addedRows)),
+		retFieldTypes:    retTypes(us),
+		outputOffset:     outputOffset,
+		handleBytes:      make([]byte, 0, 16),
+		memIdxHandles:    make(map[int64]struct{}, len(us.dirty.addedRows)),
+		belowHandleIndex: us.belowHandleIndex,
+	}
+}
+
+func (m *memIndexReader) getMemRows() ([][]types.Datum, error) {
+	tps := make([]*types.FieldType, 0, len(m.index.Columns)+1)
+	cols := m.table.Columns
+	for _, col := range m.index.Columns {
+		tps = append(tps, &cols[col.Offset].FieldType)
+	}
+	if m.table.PKIsHandle {
+		for _, col := range m.table.Columns {
+			if mysql.HasPriKeyFlag(col.Flag) {
+				tps = append(tps, &col.FieldType)
+				break
+			}
+		}
+	} else {
+		// ExtraHandle Column tp.
+		tps = append(tps, types.NewFieldType(mysql.TypeLonglong))
+	}
+
+	mutableRow := chunk.MutRowFromTypes(m.retFieldTypes)
+	err := iterTxnMemBuffer(m.ctx, m.kvRanges, func(key, value []byte) error {
+		data, err := m.decodeIndexKeyValue(key, value, tps)
+		if err != nil {
+			return err
+		}
+		handle := data[m.belowHandleIndex].GetInt64()
+		m.memIdxHandles[handle] = struct{}{}
```

Comment:
should we only insert the handle to `m.memIdxHandles` when it passed the filter on line 100?

Reply:
No, `m.memIdxHandles` is used to record the handle which is checked in `memory buffer`, not only the handle which is passed the filter.
eg: 
```go
handle in memory: 1,2
handle after passed filter in memory: 2
```
after when `getMissIndexRowsByHandle`  check handle 1, no need to get handle 1 value again in memory, because handle 1 was already checked, but not passed the filter.